DESIGN.md

```diff
@@ -0,0 +1,79 @@
+- GuestMemoryRegion: used to represent a continuous region of VM's physical memory.
+- GuestMemory: used to represent a collection of GuestMemoryRegion objects. The main responsibilities of the GuestMemory trait are:
+	- hide the detail of accessing VM's physical address (for example complex hierarchical structures).
+	- map a request address to a GuestMemoryRegion object and relay the request to it.
+	- handle cases where an access request spanning two or more GuestMemoryRegion objects.
+
+The VM memory consumers, such as virtio device drivers, vhost drivers and boot loaders etc, should only rely on traits and structs defined here to access VM's physical memory.
+
+### A Sample and Default Backend Implementation Based on mmap()
+Provide a default and sample implementation of the GuestMemory trait by mmapping VM's physical memory into current process. Three data structures are defined here:
+- MmapRegion: mmap a continous range of VM's physical memory into current and provide methods to access the mmapped memory.
+- GuestRegionMmap: a wrapper structure to map VM's physical address into (mmap\_region, offset) tuple.
+- GuestMemoryMmap: manage a collection of GuestRegionMmap objects for a VM.
+
+One of the main responsibilities of the GuestMemoryMmap object is to handle the use cases where an access request crosses the memory region boundary. This scenario may be triggered when memory hotplug is supported. So there's a tradeoff between functionality and code complexity:
+- use following pattern for simplicity which fails when the request crosses region boundary. It's current default behavior in the crosvm and firecracker project.
+```rust
+	let guest_memory_mmap: GuestMemoryMmap = ...
+	let addr: GuestAddress = ...
+        let buf = &mut [0u8; 5];
+	let result = guest_memory_mmap.find_region(addr).unwrap().write_slice(buf, addr);
+```
+- use following pattern for functionality to support request crossing region boundary:
+```rust
+	let guest_memory_mmap: GuestMemoryMmap = ...
+	let addr: GuestAddress = ...
+        let buf = &mut [0u8; 5];
+	let result = guest_memory_mmap.write_slice(buf, addr);
+```
+
+### Utilities and Helpers
+Following utility and helper traits/macros are imported from the [crosvm project](https://chromium.googlesource.com/chromiumos/platform/crosvm/) with minor changes:
+- DataInit: Types for which it is safe to initialize from raw data. A type `T` is `DataInit` if and only if it can be initialized by reading its contents from a byte array. This is generally true for all plain-old-data structs.  It is notably not true for any type that includes a reference.
+- {Le,Be}\_{16,32,64}: Explicit endian types useful for embedding in structs or reinterpreting data.
+
+## Relationship among Traits and Structs
```

Comment:
I would replace that with:

## Relationships between Traits, Structs and new Types.